src/com/palmergames/bukkit/towny/listeners/TownyPlayerListener.java

```diff
@@ -1003,6 +1003,83 @@ public void onJailedPlayerUsesCommand(PlayerCommandPreprocessEvent event) {
 		}
 	}
 	
+	/** 
+	 * Allows restricting commands while being on an town.
+	 * Also allows limiting commands to self owned plots only.
+	 * Works almost the same way as jail command blacklisting, except has more stuff
+	 * @param event PlayerCommandPreprocessEvent
+	 */
+	@EventHandler(priority = EventPriority.NORMAL)
+	public void onPlayerUsesCommandInsideTown(PlayerCommandPreprocessEvent event) {
+		if (plugin.isError() || !TownySettings.allowTownCommandBlacklisting() ||
+				!TownyAPI.getInstance().isTownyWorld(event.getPlayer().getWorld()))
+			return;
+		
+		Player player = event.getPlayer();
+		Resident res = TownyUniverse.getInstance().getResident(player.getUniqueId());
+		
+		if (!TownyAPI.getInstance().isTownyWorld(event.getPlayer().getWorld()))
+			return;
+		
+		if (res == null || TownyUniverse.getInstance().getPermissionSource().has(player, PermissionNodes.TOWNY_ADMIN_TOWN_COMMAND_BLACKLIST_BYPASS.getNode()))
+			return;
+		
+		String[] split = event.getMessage().substring(1).split(" ");
+		
+		if (TownySettings.getPlayerOwnedPlotLimitedCommands().contains(split[0])) {
+			if (TownyAPI.getInstance().isWilderness(player.getLocation())) {
+				TownyMessaging.sendErrorMsg(player, Translation.of("msg_command_limited"));
+				event.setCancelled(true);
+				return;
+			}
+			
+			WorldCoord worldCoord = WorldCoord.parseWorldCoord(player.getLocation());
+			
+			if (worldCoord.hasTownBlock()) {
+				TownBlock tb = null;
+				
+				try {
+					tb = worldCoord.getTownBlock();
+				} catch(NotRegisteredException nre) {
+					return;
+				}
+				
+				if (tb.hasResident()) {
+					Resident owner = null;
+					
+					try {
+						owner = tb.getResident();
+					} catch(NotRegisteredException nre) { return; }
+					
+					if (TownyUniverse.getInstance().getPermissionSource().has(player, "towny.claimed.owntown.*")) return;
+					
+					TownyMessaging.sendMsg("OwnerName: " + owner.getName());
+					TownyMessaging.sendMsg("PlayerName: " + player.getName());
+					
+					if (!owner.getName().equals(player.getName())) {
+						TownyMessaging.sendErrorMsg(player, Translation.of("msg_command_limited"));
+						event.setCancelled(true);
+						return;
+					} else {
+						return;
+					}
+				} else {
+					if (!TownyUniverse.getInstance().getPermissionSource().has(player, "towny.claimed.townowned.*")) {
```

Comment:
This bypass will probably also need to make sure that the `townblock.getTown().hasResident(player.getName())` otherwise you'll have mayors able to run the commands in any town.